`<vector>`: `std::vector<std::shared_ptr<mytype<T>>>` fails to compile when T is incomplete

# Describe the bug

Creating a vector of `shared_ptr` of a template instantiated incomplete type fails to compile with `type uses undefined class`

# Test case
Source:
```c++
#include <vector>
#include <memory>

template <typename T>
struct holder {
    T t;
};

class fwd;
using hptr = std::shared_ptr<holder<fwd>>;

std::vector<hptr> vec(5);
```

Output:
```
example.cpp
<source>(6): error C2079: 'holder<fwd>::t' uses undefined class 'fwd'
C:/data/msvc/14.36.32530-Pre/include\vector(2043): note: see reference to class template instantiation 'holder<fwd>' being compiled
C:/data/msvc/14.36.32530-Pre/include\vector(2033): note: while compiling class template member function 'void std::vector<hptr,std::allocator<hptr>>::_Tidy(void) noexcept'
C:/data/msvc/14.36.32530-Pre/include\vector(765): note: see the first reference to 'std::vector<hptr,std::allocator<hptr>>::_Tidy' in 'std::vector<hptr,std::allocator<hptr>>::~vector'
<source>(12): note: see the first reference to 'std::vector<hptr,std::allocator<hptr>>::~vector' in 'void __cdecl `dynamic atexit destructor for 'vec''(void)'
<source>(12): note: see reference to class template instantiation 'std::vector<hptr,std::allocator<hptr>>' being compiled
```

Godbolt: https://godbolt.org/z/738sMK337

# STL version
* visual studio 17.6.0
* reproducible with godbolt msvc latest

# Additional info

It seems the problem is in `<vector>` and not in `memory`. Instantiating `hptr` from the demo when no `std::vector` is involved compiles fine,

This code, for example, compiles:

```c++
#include <memory>

template <typename T>
struct holder {
    T t;
};

class fwd;
using hptr = std::shared_ptr<holder<fwd>>;

template <typename T>
struct my_vec {
    T* data;
    my_vec(int n) { data = new T[n]; }
    ~my_vec() { delete[] data; }
};

my_vec<hptr> vec(5);
```

Godbolt link: https://godbolt.org/z/zWd6E13Y3

This is also DevCom-10535345 / VSO-1926908 / AB#1926908.

--- a/stl/inc/vector
+++ b/stl/inc/vector
@@ -2046,7 +2046,7 @@ private:
         _My_data._Orphan_all();
 
         if (_Myfirst) { // destroy and deallocate old array
-            _Destroy_range(_Myfirst, _Mylast, _Al);
+            _STD _Destroy_range(_Myfirst, _Mylast, _Al);
             _ASAN_VECTOR_REMOVE;
             _Al.deallocate(_Myfirst, static_cast<size_type>(_Myend - _Myfirst));
 

--- a/stl/inc/xmemory
+++ b/stl/inc/xmemory
@@ -986,12 +986,12 @@ public:
     _CONSTEXPR20 void deallocate(_Ty* const _Ptr, const size_t _Count) {
         _STL_ASSERT(_Ptr != nullptr || _Count == 0, "null pointer cannot point to a block of non-zero size");
         // no overflow check on the following multiply; we assume _Allocate did that check
-        _Deallocate<_New_alignof<_Ty>>(_Ptr, sizeof(_Ty) * _Count);
+        _STD _Deallocate<_New_alignof<_Ty>>(_Ptr, sizeof(_Ty) * _Count);
     }
 
     _NODISCARD_RAW_PTR_ALLOC _CONSTEXPR20 __declspec(allocator) _Ty* allocate(_CRT_GUARDOVERFLOW const size_t _Count) {
         static_assert(sizeof(value_type) > 0, "value_type must be complete before calling allocate.");
-        return static_cast<_Ty*>(_Allocate<_New_alignof<_Ty>>(_Get_size_of_n<sizeof(_Ty)>(_Count)));
+        return static_cast<_Ty*>(_STD _Allocate<_New_alignof<_Ty>>(_Get_size_of_n<sizeof(_Ty)>(_Count)));
     }
 
 #if _HAS_CXX23
@@ -1104,7 +1104,7 @@ _CONSTEXPR20 void _Destroy_range(_Alloc_ptr_t<_Alloc> _First, const _Alloc_ptr_t
     using _Ty = typename _Alloc::value_type;
     if constexpr (!conjunction_v<is_trivially_destructible<_Ty>, _Uses_default_destroy<_Alloc, _Ty*>>) {
         for (; _First != _Last; ++_First) {
-            allocator_traits<_Alloc>::destroy(_Al, _Unfancy(_First));
+            allocator_traits<_Alloc>::destroy(_Al, _STD _Unfancy(_First));
         }
     }
 }
@@ -1114,7 +1114,7 @@ _CONSTEXPR20 void _Destroy_range(_NoThrowFwdIt _First, const _NoThrowSentinel _L
     // note that this is an optimization for debug mode codegen; in release mode the BE removes all of this
     if constexpr (!is_trivially_destructible_v<_Iter_value_t<_NoThrowFwdIt>>) {
         for (; _First != _Last; ++_First) {
-            _Destroy_in_place(*_First);
+            _STD _Destroy_in_place(*_First);
         }
     }
 }

--- a/tests/libcxx/expected_results.txt
+++ b/tests/libcxx/expected_results.txt
@@ -913,11 +913,6 @@ std/re/re.alg/re.alg.search/awk.locale.pass.cpp FAIL
 std/re/re.alg/re.alg.search/basic.locale.pass.cpp FAIL
 std/re/re.alg/re.alg.search/ecma.locale.pass.cpp FAIL
 std/re/re.alg/re.alg.search/extended.locale.pass.cpp FAIL
-
-# Not analyzed. Involves incomplete types.
-std/utilities/memory/allocator.traits/allocator.traits.members/construct.pass.cpp FAIL
-std/utilities/memory/allocator.traits/allocator.traits.members/destroy.pass.cpp FAIL
-
 # Not analyzed. Error mentions allocator<const T>.
 std/utilities/memory/specialized.algorithms/specialized.construct/construct_at.pass.cpp FAIL
 

--- a/tests/std/tests/VSO_0000000_instantiate_containers/test.compile.pass.cpp
+++ b/tests/std/tests/VSO_0000000_instantiate_containers/test.compile.pass.cpp
@@ -25,6 +25,7 @@
 #include <iterator>
 #include <list>
 #include <map>
+#include <memory>
 #include <numeric>
 #include <queue>
 #include <random>
@@ -103,6 +104,8 @@ struct incomplete;
 using validating_falsity   = tagged_falsity<holder<incomplete>>;
 using validating_less      = tagged_less<holder<incomplete>>;
 using validating_converter = tagged_converter<holder<incomplete>>;
+
+vector<shared_ptr<holder<incomplete>*>> unused_pointers_to_incompletable;
 #endif // ^^^ no workaround ^^^
 
 template <typename T, typename U>

